blocks/api/children.js

```diff
@@ -0,0 +1,132 @@
+/**
+ * External dependencies
+ */
+import { castArray } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { renderToString } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import * as node from './node';
+
+/**
+ * A representation of a block's rich text value.
+ *
+ * @typedef {WPBlockChild[]} WPBlockChildren
+ */
+
+/**
+ * Given a block node, returns a serialize-capable WordPress element.
+ *
+ * @param {WPBlockChildren} children Block node to convert.
+ *
+ * @return {WPElement} A serialize-capable element.
+ */
+export function getSerializeCapableElement( children ) {
+	// The fact that a block node is compatible with the element serializer is
+	// merely an implementation detail that currently serves to be true, but
+	// should not be mistaken as being a guarantee on the external API. The
+	// public API only offers guarantees to work with strings (toHTML) and DOM
+	// elements (fromDOM), and should provide utilities to manipulate the value
+	// rather than expect consumers to inspect or construct its shape (concat).
+	return children;
+}
+
+/**
+ * Given two or more block nodes, returns a new block node representing a
+ * concatenation of its values.
+ *
+ * @param {...WPBlockChildren} blockNodes Block nodes to concatenate.
+ *
+ * @return {WPBlockChildren} Concatenated block node.
+ */
+export function concat( ...blockNodes ) {
+	const result = [];
+	for ( let i = 0; i < blockNodes.length; i++ ) {
+		const blockNode = castArray( blockNodes[ i ] );
+		for ( let j = 0; j < blockNode.length; j++ ) {
+			const child = blockNode[ j ];
+			const canConcatToPreviousString = (
+				typeof child === 'string' &&
+				typeof result[ result.length - 1 ] === 'string'
+			);
+
+			if ( canConcatToPreviousString ) {
+				result[ result.length - 1 ] += child;
+			} else {
+				result.push( child );
+			}
+		}
+	}
+
+	return result;
+}
+
+/**
+ * Given an iterable set of DOM nodes, returns equivalent block children.
+ * Ignores any non-element/text nodes included in set.
+ *
+ * @param {Iterable.<Node>} domNodes Iterable set of DOM nodes to convert.
+ *
+ * @return {WPBlockChildren} Block children equivalent to DOM nodes.
+ */
+export function fromDOM( domNodes ) {
+	const result = [];
+	for ( let i = 0; i < domNodes.length; i++ ) {
+		try {
+			result.push( node.fromDOM( domNodes[ i ] ) );
+		} catch ( error ) {
+			// Simply ignore if DOM node could not be converted.
```

Comment:
When can this happen?

Reply:
>When can this happen?

The `child` matcher throws on anything other than `nodeType` of `TEXT_NODE` or `ELEMENT_NODE`, so if a block uses a `children` selector on an element which includes a comment node, for example, this would prevent its inclusion.

```html
<div class="my-block"><!-- foo -->hello world</div>
```

```json
{
  "type": "array",
  "source": "children",
  "selector": "div"
},
```

https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Constants